simulators/main/event.py

```diff
@@ -0,0 +1,318 @@
+            else:
+                log.debug(f'{self.device_id} followedBy task missed. probability = {1 - followed_by_frequency / this_frequency} , def: {json.dumps(followed_by_definition)}')
+
+    def create_one_time_task(self, event_definition, start_in_seconds=2, event_callback=None):
+        callback = event_callback or Event.event_mapping[event_definition["type"]]
+        task = Task(start_in_seconds, lambda task: {self.execute_callback_and_remove_task(callback, event_definition, task)})
+        return task
+
+    def execute_callback_and_remove_task(self, callback, event_definition, task):
+        callback(self, event_definition, task)
+        if task in self.tasks:
+            self.tasks.remove(task)
+            log.debug(f'{self.device_id} task({id(task)}) removed: {json.dumps(event_definition)}, tasks: {len(self.tasks)}')
+
+    def send_event(self, event_fragment, timestamp=None):
+        newTimestamp = timestamp or interface.current_timestamp()
+        base_event = {
+            'source': {
+                'id': self.device_id
+            },
+            'time': newTimestamp
+        }
+        base_event.update(event_fragment)
+
+        if self.shutdown:
+            log.info(f'{self.model["id"]} is down -> ignore event: {json.dumps(base_event)}')
+            return None
+        else:
+            response = cumulocityAPI.send_event(base_event)
+            if response:
+                log.info(f"Created new {event_fragment.get('type')} event for device {self.model.get('label')}, id {self.model.get('id')}")
+            return newTimestamp
+
+    def is_in_productionTime(self):
+        # if there are no shiftplans for a device, it should not be affected by production-time
+        has_no_shiftplan = True
+        for shiftplan in self.shiftplans:
+            if shiftplan.locationId == self.locationId:
+                has_no_shiftplan = False
+                for timeslot in shiftplan.recurringTimeSlots:
+                    if timeslot.slotType == "PRODUCTION":
+                        now = datetime.utcnow()
+                        start = datetime.strptime(timeslot.slotStart, interface.DATE_FORMAT)
+                        end = datetime.strptime(timeslot.slotEnd, interface.DATE_FORMAT)
+                        if start < now < end:
+                            return True
+        return has_no_shiftplan
+
+    event_mapping = {'Availability': on_availability_event,
+                     'Piece_Produced': on_piece_produced_event,
+                     'Pieces_Produced': on_pieces_produced_event,
+                     'Piece_Ok': on_piece_ok_event,
+                     'Pieces_Ok': on_pieces_ok_event,
+                     'Piece_Quality': on_piece_quality_event,
+                     'Shutdown': on_shutdown_event}
+
+    def tick(self):
+        if not self.enabled:
```

Comment:
You could solve this + check for EventInProduction by using a method. Have a implementation for both and in the MachineSimulator you would have your tick function which just calls the method and then ticks depending on the return value of the function (true/false)